core: make idle thread optional

kaspar030 opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

This PR makes the idle thread optional, if the architecture implements support for that.
In order to do that, an architecture needs to implement `sched_arch_idle()` and disable the module `core_idle_thread`.

The main benefits are:

- saving a full thread context and stack
- faster and shorter sleep transitions (no need to schedule idle thread first)

The PR includes an implementation for CortexM3 and higher (M0+ doesn't support dynamically changing IRQ priorities, which the current implementation relies on).

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure

Any threading application obviously still needs to work. Also, the pm machinery needs to work as before.

Any shell test should confirm that the scheduler is idling (not switching to any thread if there's no runnable) and reacting to ISR (react to uart input and re-schedule shell thread).

I tested power management using tests/periph_pm, confirming that unblocking the right mode for a while makes the shell unresponsive (power mode has been set), and that when the RTC interrupt triggers, shell works again.

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->

*edit* the commit history looks a bit scary due to #13825 being included. The new logic of this PR is contained in the latest two commits (one for basic support in core, one for the cortexm implementation).

### Issues/PRs references

This PR needs #13825, which re-writes the Cortex-M PendSV logic so it can handle the case where there's no thread to save on entry.

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

bergzand commented:
This breaks the schedstatistics module :cry:

maribu commented:
> disable the module `core_idle_thread`.

I would like to inverse the logic. Could we instead have a feature (and corresponding module) such as `thread_less_idling`, which would always be in `FEATURS_OPTIONAL`?

Providing a feature instead of disabling a module seems more natural to me.

kaspar030 commented:
> Providing a feature instead of disabling a module seems more natural to me.

done

kaspar030 commented:
> This breaks the schedstatistics module cry

This'll need reworking the sched_cb logic: `sched_cb(sched_active_pid, next_thread->pid);` expects ending of one and starting the next thread to happen at the same time.

kaspar030 commented on core/sched.c:
```diff
@@ -80,11 +80,14 @@ int __attribute__((used)) sched_run(void)
     sched_context_switch_request = 0;
 
+#ifndef MODULE_CORE_IDLE_THREAD
+    while (!runqueue_bitcache) {
```
note to self: as is, the compiler must assume that sched_arch_idle() calls a function that modifies runqueue_bitcache (it is static in this compilation unit). With LTO, it might come to a different result. need to make this atomic or volatile.

kaspar030 commented:
I marked this WIP, as there's some code that expects that there's always an active thread (e.g., most of the IPC code due to sched_switch()).

fjmolinas commented:
this one needs a rebase now :)

kaspar030 commented:
> I marked this WIP, as there's some code that expects that there's always an active thread (e.g., most of the IPC code due to sched_switch()).

It seems to me that all that code is actually called when within thread context, meaning that having no active thread won't happen. So I've removed WIP.

kaspar030 commented:
@leandrolanzieri I need some suggestions on where to put the new "no_idle_thread" feature.

leandrolanzieri commented:
> @leandrolanzieri I need some suggestions on where to put the new "no_idle_thread" feature.

I think that the definition of `HAS_NO_IDLE_THREAD` can be placed in `/kconfigs/Kconfig.features`.

leandrolanzieri commented on cpu/cortexm_common/Makefile.features:
```diff
@@ -30,3 +30,8 @@ else ifeq ($(CPU_CORE),cortex-m23)
 endif
+
+# cortex-m3 and higher don't need the idle thread
+ifneq (,$(filter cortex-m2% cortex-m4% cortex-m3% cortex-m7%,$(CPU_ARCH)))
```
```suggestion
ifneq (,$(filter cortex-m2% cortex-m4% cortex-m3% cortex-m7%,$(CPU_CORE)))
```

leandrolanzieri replied:
On second thoughts, maybe checking for the architecture is cleaner?
```Makefile
ifneq (,$(filter armv7 armv8,$(CPU_ARCH)))
```

kaspar030 replied:
I don't mind, personally I'm thinking "cortex-m3 and higher" instead of "armv7". your call.

leandrolanzieri replied:
I would go with the arch approach just to be in line with what you added in the cortexm_common Kconfig.

kaspar030 replied:
ok

leandrolanzieri replied:
Oh I think the correct values are `armv7m` and `armv8m` :/

kaspar030 replied:
:) fixed

kaspar030 commented:
We had some offline discussions on this PR. I checked the core code, this shouldn't(tm) cause issues. @fjmolinas is running a full test run on iotlab-m3. I think, should that not cause any issues, we should merge this, so it get's some use over the next weeks before release.

Should this cause any non-fixable issues (or any trouble during release testing), this can be disabled by removing the "FEATURES_OPTIONAL" that pulls it in in the first place.

@miri64 does that sound sensible to you?

Just to assess the impact, some parts of the code are slightly changed for all platforms (sched_cb handling KERNEL_PID_UNDEF, and factoring out `_unschedule()` in `sched_run()`).
The jucy bits are only enabled on Cortex-M3 and higher.

miri64 commented:
> Should this cause any non-fixable issues (or any trouble during release testing), this can be disabled by removing the "FEATURES_OPTIONAL" that pulls it in in the first place.
> 
> @miri64 does that sound sensible to you?

Yes. I would however still insist on not merging this after soft feature freeze.

kaspar030 commented:
> Yes. I would however still insist on not merging this after soft feature freeze.

Agreed, this is under the assumption that it gets ACK^2'ed this week.

fjmolinas commented:
I launched the tests in  `04-single-hop-6lowpan-icmp` that target `iotlab-m3` boards so 1-4 and 9-10. All tests where passing.

kaspar030 commented:
I pushed a fix for tests/shell.
(and I accidentally rebased while doing so)

bergzand commented on sys/schedstatistics/schedstatistics.c:
```diff
@@ -31,15 +31,18 @@ void sched_statistics_cb(kernel_pid_t active_thread, kernel_pid_t next_thread)
-    schedstat_t *active_stat = &sched_pidlist[active_thread];
-    active_stat->runtime_ticks += now - active_stat->laststart;
+    /* Update active thread stats */
+    if (active_thread != KERNEL_PID_UNDEF) {
```
Should it also be documented in the SCHED_CB docs that `KERNEL_PID_UNDEF` is possible now?

kaspar030 replied:
there's not much documentation for sched_cb I think, other than that the whole module does what it does. where would you like to see extra documentation?

bergzand replied:
Maybe leave this to a follow-up, there is indeed not much documentation, better do this right in a new PR, I can open something later this week I think

bergzand replied:
See #14334

kaspar030 commented on core/init.c:
```diff
@@ -64,17 +67,19 @@ static void *idle_thread(void *arg)
 {
     irq_disable();
 
+#ifdef MODULE_CORE_IDLE_THREAD
```
sched_arch_idle() would be undefined in that case?

bergzand replied:
But that doesn't matter for this file right?

miri64 replied:
`IS_USED` ;-)

kaspar030 replied:
changed to IS_USED, here and in sched.c

kaspar030 commented on core/sched.c:
nice one, the function call is missing there...

kaspar030 replied:
I guess we have to re-run the tests.

fjmolinas replied:
Will do

fjmolinas replied:
Re-tested, all is still good.

kaspar030 commented:
- rebased to include the sched_cb doc change (which is actually inaccurate without this PR)
- changed to use IS_USED instead of ifdefs. did some comparison on samr21-xpro (which does use the idle thread), no size diff, so I'm assuming the linker does its job.

bergzand commented on tests/shell/tests/01-run.py:
```diff
@@ -60,6 +59,8 @@
     ('end_test', '[TEST_END]'),
 )
 
+CMDS_REGEX = { 'ps' }
```
Flake8 wants it like this:
```suggestion
CMDS_REGEX = {'ps'}
```

bergzand replied:
Feel free to amend this directly

fjmolinas replied:
I foresee a conflict with #13197, as a headsup.

kaspar030 replied:
seems like git manages to merge this just fine. :smile:

kaspar030 replied:
fixed

bergzand commented:
Please Squash!

kaspar030 commented:
squashed

fjmolinas commented on core/include/sched.h:
```diff
@@ -215,13 +229,6 @@ NORETURN void sched_task_exit(void);
  */
 typedef void (*sched_callback_t)(kernel_pid_t active, kernel_pid_t next);
-
-/**
- * @brief  Register a callback that will be called on every scheduler run
- *
- * @param[in] callback The callback functions that will be called
- */
-void sched_register_cb(sched_callback_t callback);
```
Was this removal intentional?

kaspar030 replied:
nope, I put it back and amended

fjmolinas commented:
I believe failing tests are due to a faulty worker. @kaspar030 anyway that could take a look at FU?

fjmolinas commented:
These are the failing tests, I'll re-run locally"

<details>

```
    - tests/driver_apds99xx_full
    - tests/cond_order
    - tests/driver_apds99xx
    - tests/evtimer_msg
    - tests/devfs
    - examples/micropython
    - tests/pkg_libb2
    - tests/pkg_tinycbor
    - tests/bench_runtime_coreapis
    - tests/pkg_nanopb
    - tests/progress_bar
    - tests/test_tools
    - tests/pthread_condition_variable
    - tests/pkg_ucglib
    - tests/pkg_yxml
    - tests/rmutex_cpp
    - tests/posix_time
    - tests/trace
    - tests/periph_hwrng
    - tests/trickle
    - tests/ps_schedstatistics
    - tests/rmutex
    - tests/isr_yield_higher
    - tests/pkg_umorse
    - tests/xtimer_usleep
    - tests/gnrc_ndp
    - tests/xtimer_remove
    - tests/xtimer_usleep_short
    - tests/heap_cmd
    - tests/thread_flags
    - tests/shell
    - tests/thread_cooperation
    - tests/thread_race
    - tests/thread_msg_bus
    - tests/pkg_spiffs
    - tests/ssp
    - tests/sched_testing
    - tests/thread_exit
    - tests/thread_msg_block_wo_queue
    - tests/xtimer_hang
    - tests/thread_msg
    - tests/pipe
    - tests/vfs_plus_stdio
    - tests/xtimer_now32_overflow
    - tests/unittests
    - tests/pkg_monocypher
    - tests/pthread_barrier
    - tests/cpp_ext
    - tests/cpp11_thread
    - tests/bloom_bytes
```
</details>

kaspar030 commented:
> I believe failing tests are due to a faulty worker. @kaspar030 anyway that could take a look at FU?

I've restarted the usual suspect, let me re-trigger here.

kaspar030 commented:
Turns out the unittest hard fault was caused by too a too small stack for main. I've added a commit to increase it slightly. (stack is also small on master, depending on compiler, it hard faults there without increase. This was unnoticed as the crash happens after finishing the tests).

miri64 commented:
This PR still needs two ACKs to be merged before soft feature freeze (which I will anounce in a few hours) ;-).

bergzand commented:
And go!

kaspar030 commented:
> And go!

Thanks for reviewing!

miri64 commented:
Now let's rethink making the [`main` thread optional](https://github.com/RIOT-OS/RIOT/pull/10971) ;-P

fjmolinas commented:
Bisecting I have 0ff9e554eb712fe4f37ea46e232eaed9c20b9071 causing `tests/xtimer_periodic_wakeup` to get stuck.

```
Testing interval 25... (now=4054509)
Testing interval 24... (now=4057901)
Testing interval 23... (now=4061293)
Testing interval 22... (now=4064686)
Testing interval 21... (now=4068078)
Testing interval 20... (now=4071461)
Testing interval 19... (now=4074844)
Testing interval 18... (now=4078229)
Testing interval 17... (now=4081612)
Testing interval 16... (now=4084996)
# gets stuck here
```

bergzand commented:
Totally not related to #14557, some measurements on the same54-xpro with bench_mutex_pingpong:

| | idle thread | no_idle_thread | impact|
|-|-|-|-|
| GCC | 262582 | 251572 | 4.1% |
| GCC+LTO | 272109 | 252631 | 7.1% |
|LLVM | 260303 | 244897 | 5.9% |
